### PR TITLE
mitmproxy: 9.0.1 -> 10.1.1

### PR DESCRIPTION
**Pull request by K900**
Also add the new mitmproxy-rs dependency, which seemingly replaces mitmproxy-wireguard.

## Description of changes

Got nerdsniped into it on Matrix. @fabaff I hope you don't mind being listed as maintainer for the Rust bits.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 264207` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>mitmproxy (python311Packages.mitmproxy)</li>
    <li>mitmproxy.dist (python311Packages.mitmproxy.dist)</li>
    <li>mitmproxy2swagger</li>
    <li>mitmproxy2swagger.dist</li>
    <li>python310Packages.mitmproxy</li>
    <li>python310Packages.mitmproxy-rs</li>
    <li>python310Packages.mitmproxy-rs.dist</li>
    <li>python310Packages.mitmproxy.dist</li>
    <li>python311Packages.mitmproxy-rs</li>
    <li>python311Packages.mitmproxy-rs.dist</li>
    <li>wapiti</li>
    <li>wapiti.dist</li>
  </ul>
</details>

**Comment by K900** [comment]
Pretty sure this is nixpkgs-review messing with the PYTHONPATH. I wonder if we should just make `buildPythonApplication` drop outside PYTHONPATH unconditionally...

**Review by Ma27 — APPROVED** [review]
Works fine for me :+1:

**Comment by K900** [comment]
I'm just going to merge this before it bitrots.

File: pkgs/development/python-modules/mitmproxy-rs/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{ lib
+, stdenv
+, buildPythonPackage
+, fetchFromGitHub
+, rustPlatform
+, darwin
+}:
+
+buildPythonPackage rec {
+  pname = "mitmproxy-rs";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mitmproxy-rs/default.nix:10]: If we go really strongly by the contributing guide then this should have been it's own commit but it's alright for me if I am honest :)

File: pkgs/development/python-modules/mitmproxy-rs/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  pythonImportsCheck = [ "mitmproxy_rs" ];
+
+  meta = with lib; {
+    description = "The Rust bits in mitmproxy";
+    homepage = " https://github.com/mitmproxy/mitmproxy_rs";
+    changelog = "https://github.com/mitmproxy/mitmproxy_rs/blob/${src.rev}/CHANGELOG.md";
+    license = licenses.mit;
+    maintainers = with maintainers; [ fab ];
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mitmproxy-rs/default.nix:47]: We default to unix here by buildPythonPackage which is more correct than all

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -6856,6 +6856,8 @@ self: super: with self; {
 
   mitmproxy = callPackage ../development/python-modules/mitmproxy { };
 
+  mitmproxy-rs = callPackage ../development/python-modules/mitmproxy-rs { };
+
   mitmproxy-wireguard = callPackage ../development/python-modules/mitmproxy-wireguard { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:6861]: Dropping that in another PR. To quote the PR:

https://togithub.com/mitmproxy/mitmproxy/pull/5909

> mitmproxy-wireguard is already shipping in 9.0, but we will add more non-WireGuard stuff written in Rust.

**Pull request merged**